### PR TITLE
Housing and Urban Development (improved HUD)

### LINKED ISSUE #309: Key colors
As a player, I want to have more colors for keys, so that I can open more doors.

Improve the HUD to support more key colors.

### LINKED ISSUE #311: Bomb improvements - Bomb inventory slot
Create a separate inventory slot that only holds one bomb.

This is a subtask of #310 - Bomb improvements

### LINKED ISSUE #312: Bomb improvements - Improve timer
Make the timers start when the bomb is picked up instead of when it is placed down.

This is a subtask of #310 - Bomb improvements

### PR DESCRIPTION
**Pull request by ChielBruin**
Improved the HUD class to be more organized and flexible.
- More flexible key display
  - Allows more than 3 keys
  - Allows more than 3 key colors
- Better heart display
  - Allows as much health as you want
  - Works with health regeneration
  - Works with non-full health at the start
- Testing
  - 91% testing coverage of the HUD

This PR also includes the addition of TickProducers that decrease code reuse.

Closes #309, closes #311 and closes #312

### CODE REVIEW

**Comment by ChielBruin** [comment]
Some things are still missing, so I will close this PR for now.

**Comment by Geweldig** [comment]
By the looks of it this also contains Wytze's branch which has not been merged yet, is this supposed to be the case?

**Comment by ChielBruin** [comment]
Also added the game timer, improved bomb display and explosions moved some positions.

Closes #311, closes #312 
This pull request should not be merged before #331

File: src/main/java/nl/tudelft/contextproject/Main.java [hunk 1]
```diff
@@ -427,4 +411,11 @@ public BitmapFont getGuiFont() {
 	public Controller getController() {
 		return controller;
 	}
+	
+	/**
+	 * Loads the gui font.
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/Main.java]: Loads and sets the GUI font.

File: src/main/java/nl/tudelft/contextproject/controller/GameController.java [hunk 2]
```diff
@@ -246,4 +248,15 @@ public void gameEnded(boolean didElvesWin) {
 		Main main = Main.getInstance();
 		main.setController(new EndingController(main, didElvesWin));
 	}
+
+	/**
+	 * Method used for testing.
+	 * sets the hud to a new hud.
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/controller/GameController.java]: "sets the hud to a new hud" can be removed, as that should be the case with a setter

File: src/main/java/nl/tudelft/contextproject/hud/HUD.java [hunk 3]
```diff
@@ -35,154 +42,237 @@
...
+		for (int i = 0; i < VRPlayer.PLAYER_MAX_HEALTH; i++) {
+			heartContainer.attachChild(getHealthContainer(i));
+		}
+		updateHearts(Main.getInstance().getCurrentGame().getPlayer());
+	}
+
+	/**
+	 * Attaches the bomb icon and counter to the HUD.
+	 * 
+	 * @param b
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/hud/HUD.java]: "b" could do with a more descriptive name

File: src/main/java/nl/tudelft/contextproject/hud/HUD.java [hunk 4]
```diff
@@ -35,154 +42,237 @@
...
+	 * @param container
+	 * 		the new container where keys will be stored
+	 */
+	protected void setKeyContainer(Node container) {
+		keyContainer = container;
+	}
+
+	/**
+	 * Method used for testing.
+	 * Set the bomb node to a custom node.
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/hud/HUD.java]: Same comment as before.

File: src/main/java/nl/tudelft/contextproject/hud/HUD.java [hunk 5]
```diff
@@ -35,154 +42,237 @@
...
+		List<ColorRGBA> keys = inventory.getKeyColors();
+		for (ColorRGBA c : keys) {
+			keyContainer.attachChild(getKeyImage(keys.size(), i, c));
+			i++;
 		}
 	}
+	
+	/**
+	 * Method used for testing.
+	 * Sets the key container to a custom container.
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/hud/HUD.java]: Same comment as before.

File: src/main/java/nl/tudelft/contextproject/hud/HUD.java [hunk 6]
```diff
@@ -35,154 +42,237 @@
...
+	 * @param node
+	 * 		the node to replace the bomb node
+	 */
+	protected void setBombNode(Node node) {
+		bombNode = node;
+	}
+
+	/**
+	 * Method used for testing.
+	 * Set the timer text to a custom text.
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/hud/HUD.java]: Same comment as before.

File: src/main/java/nl/tudelft/contextproject/hud/HUD.java [hunk 7]
```diff
@@ -35,154 +42,237 @@
...
+	 * 
+	 * @param m
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/hud/HUD.java]: This param name makes no sense in my opinion.

File: src/main/java/nl/tudelft/contextproject/model/Inventory.java [hunk 8]
```diff
@@ -11,18 +11,19 @@
...
 
 	/**
 	 * Constructor for the inventory, starts empty with 0 keys and doors.
 	 */
 	public Inventory() {
-		this.pickedUpEntities = new ArrayList<>();
-		this.keys = 0;
-		this.bombs = 0;
+		this.keys = new ArrayList<>();
+		this.bomb = null;
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/Inventory.java]: not needed, this.bomb will be null by default.

File: src/main/java/nl/tudelft/contextproject/model/TickProducer.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Interface for classes that produce ticks that can be listened to.
+ */
+public interface TickProducer {
+
+	/**
+	 * @return
+	 * 		a list with all the attached tickListeners
+	 */
+	public List<TickListener> getTickListeners();
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/TickProducer.java]: Remove "public" to stay consistent with other interfaces.

File: src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java [hunk 10]
```diff
@@ -205,10 +212,14 @@ public CharacterControl getPhysicsObject() {
...
-		if (inventory.containsBomb()) {
-			inventory.getBomb().setPickedup(false);
-			inventory.remove(inventory.getBomb());
-		}
+		if (!inventory.containsBomb()) return;
+		
+		Bomb bomb = inventory.getBomb();
+		inventory.remove(bomb);
+		bomb.setPickedup(false);
+//		bomb.setState(EntityState.NEW);
```
Comment (Geweldig) [inline src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java]: Is this code still needed?

**Comment by Geweldig** [comment]
After my issues have been addressed I think this can be merged.

**Comment by Geweldig** [comment]
Can be merged now in my opinion.

File: src/main/java/nl/tudelft/contextproject/model/Inventory.java [hunk 11]
```diff
@@ -72,30 +66,33 @@ public void add(Bomb bomb) {
...
-				if (key.getColor().equals(toRemove.getColor())) {
-					it.remove();
-					keys--;
-					return;
-				}
-			}
+		if (ent instanceof Bomb && bomb != null) {
+			bomb = null;
+			updateTickListeners();
+			return;
```
Comment (Taeir) [inline src/main/java/nl/tudelft/contextproject/model/Inventory.java]: Remove return, add else if

File: src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java [hunk 12]
```diff
@@ -219,28 +227,20 @@ public void pickUp() {
...
-						((Bomb) ent).activate();
-						((Bomb) ent).setPickedup(true);
-						return;
-					}
-				}
-			}
-			if (ent instanceof Key) {
+			if (ent instanceof Bomb && !inventory.containsBomb()) {
+				inventory.add((Bomb) ent);
+				return;
```
Comment (Taeir) [inline src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java:234]: Add `((Bomb) ent).activate();`

File: src/test/java/nl/tudelft/contextproject/model/TickProducerTest.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+ */
+public class TickProducerTest {
+	
+	TickProducer tp;
+
+	/**
+	 * Create a fresh instance of {@link TickProducer} for each test.
+	 */
+	@Before
+	public void test() {
```
Comment (Taeir) [inline src/test/java/nl/tudelft/contextproject/model/TickProducerTest.java]: Rename to setUp

File: src/test/java/nl/tudelft/contextproject/model/TickProducerTest.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Test class for {@link TickProducer}.
+ */
+public class TickProducerTest {
+	
+	TickProducer tp;
```
Comment (Taeir) [inline src/test/java/nl/tudelft/contextproject/model/TickProducerTest.java]: Make private

File: src/test/java/nl/tudelft/contextproject/model/TickProducerTest.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Test class for {@link TickProducer}.
+ */
+public class TickProducerTest {
```
Comment (Taeir) [inline src/test/java/nl/tudelft/contextproject/model/TickProducerTest.java]: Extend TestBase

**Comment by Taeir** [comment]
Changes look good, can be merged when my comments have been addressed.

File: src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java (hunk 12 above)
Comment (ChielBruin) [inline src/main/java/nl/tudelft/contextproject/model/entities/VRPlayer.java:234]: This is fixed in #338.

**Comment by Geweldig** [comment]
If @Taeir gives the go ahead this can be merged.

**Comment by Taeir** [comment]
Looks good now, merging it.

**Pull request merged**